src/server/conn/auto.rs

```diff
@@ -116,57 +118,180 @@ enum Version {
     H1,
     H2,
 }
-async fn read_version<'a, A>(mut reader: A) -> IoResult<(Version, Rewind<A>)>
+
+fn read_version<I>(io: I) -> ReadVersion<I>
 where
-    A: Read + Unpin,
+    I: Read + Unpin,
 {
-    use std::mem::MaybeUninit;
-
-    let mut buf = [MaybeUninit::uninit(); 24];
-    let (version, buf) = ReadVersion {
-        reader: &mut reader,
-        buf: ReadBuf::uninit(&mut buf),
+    ReadVersion {
+        io: Some(io),
+        buf: [MaybeUninit::uninit(); 24],
+        filled: 0,
         version: Version::H1,
         _pin: PhantomPinned,
     }
-    .await?;
-    Ok((version, Rewind::new_buffered(reader, Bytes::from(buf))))
 }
+
 pin_project! {
-    struct ReadVersion<'a, A: ?Sized> {
-        reader: &'a mut A,
-        buf: ReadBuf<'a>,
+    struct ReadVersion<I> {
```

Comment:
Had to remove all references from this type since we can't use async/await anymore because we need a nameable `Connection` type to add the `graceful_shutdown` method to.